# Data Analytics using Python libraries

Python is a general purpose language and is often used for things other than data analysis and data science. What makes Python extremely useful for working with data?

There are libraries that give users the necessary functionality when crunching data. Below are the major Python libraries that are used for working with data. You should take some time to familiarize yourself with the basic purposes of these packages.


## PYTHON LIBRARIES FOR DATA ANALYSIS
    Numpy and Scipy – Fundamental Scientific Computing
    Pandas – Data Manipulation and Analysis
    Matplotlib – Plotting and Visualization
    


## NumPy

NumPy stands for ‘Numerical Python’ or ‘Numeric Python’. It is an open source module of Python which provides fast mathematical computation on arrays and matrices. 

NumPy provides the essential multi-dimensional array-oriented computing functionalities designed for high-level mathematical functions and scientific computation.

There are several ways to create an array in NumPy like np.array, np.zeros, no.ones, etc. Each of them provides some flexibility.

Some of the important attributes of a NumPy object are:

        Ndim: displays the dimension of the array
        Shape: returns a tuple of integers indicating the size of the array
        Size: returns the total number of elements in the NumPy array
        Dtype: returns the type of elements in the array, i.e., int64, character
        Itemsize: returns the size in bytes of each item
        Reshape: Reshapes the NumPy array

NumPy array elements can be accessed using indexing. Below are some of the useful examples:

        A[2:5] will print items 2 to 4. Index in NumPy arrays starts from 0
        A[2::2] will print items 2 to end skipping 2 items
        A[::-1] will print the array in the reverse order
        A[1:] will print from row 1 to end
        
The session covers these and some important attributes of the NumPy array object in detail.

## Pandas

Similar to NumPy, Pandas is one of the most widely used python libraries in data science. It provides high-performance, easy to use structures and data analysis tools. Unlike NumPy library which provides objects for multi-dimensional arrays, Pandas provides in-memory 2d table object called Dataframe

Hence, with 2d tables, pandas is capable of providing many additional functionalities like creating pivot tables, computing columns based on other columns and plotting graphs.

Pandas Series object is created using pd.Series function. Each row is provided with an index and by defaults is assigned numerical values starting from 0.

Some commonly used data structures in pandas are:

        Series objects: 1D array, similar to a column in a spreadsheet
        DataFrame objects: 2D table, similar to a spreadsheet
        Panel objects: Dictionary of DataFrames, similar to sheet in MS Excel

Dataframes can also be easily exported and imported from CSV, Excel, JSON, HTML and SQL database. Some other essential methods that are present in dataframes are:

        head(): returns the top 5 rows in the dataframe object
        tail(): returns the bottom 5 rows in the dataframe
        info(): prints the summary of the dataframe
        describe(): gives a nice overview of the main aggregated values over each column

#### Libraries and Imports

Once you’ve installed the modules, use the import statement to make the modules available in your program :


In [3]:
import numpy as np
import pandas as pd
pd.set_option("display.precision", 2)

#### Case Study
We'll demonstrate the main methods in action by analyzing a dataset on the churn rate of telecom operator clients. Let's read the data (using read_csv), and take a look at the first 5 lines using the head method:

In [4]:
df = pd.read_csv('C:\\Users\SK050583\Downloads\churn-in-telecoms-dataset.csv')
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [ ]:
Let’s have a look at data dimensionality, feature names, and feature types.

In [5]:
print(df.shape)

(3333, 21)


From the output, we can see that the table contains 3333 rows and 20 columns.

Now let's try printing out column names using columns:

In [6]:
print(df.columns)

Index(['state', 'account length', 'area code', 'phone number',
       'international plan', 'voice mail plan', 'number vmail messages',
       'total day minutes', 'total day calls', 'total day charge',
       'total eve minutes', 'total eve calls', 'total eve charge',
       'total night minutes', 'total night calls', 'total night charge',
       'total intl minutes', 'total intl calls', 'total intl charge',
       'customer service calls', 'churn'],
      dtype='object')


In [ ]:
We can use the info() method to output some general information about the dataframe:

In [7]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

bool, int64, float64 and object are the data types of our features. We see that one feature is logical (bool), 3 features are of type object, and 16 features are numeric. With this same method, we can easily see if there are any missing values. Here, there are none because each column contains 3333 observations, the same number of rows we saw before with shape.

We can change the column type with the astype method. Let's apply this method to the Churn feature to convert it into int64:

In [8]:
df['churn'] = df['churn'].astype('int64')

The describe method shows basic statistical characteristics of each numerical feature (int64 and float64 types): number of non-missing values, mean, standard deviation, range, median, 0.25 and 0.75 quartiles.

In [20]:
df.describe()

account length  area code  number vmail messages  total day minutes  \
count         3333.00    3333.00                3333.00            3333.00   
mean           101.06     437.18                   8.10             179.78   
std             39.82      42.37                  13.69              54.47   
min              1.00     408.00                   0.00               0.00   
25%             74.00     408.00                   0.00             143.70   
50%            101.00     415.00                   0.00             179.40   
75%            127.00     510.00                  20.00             216.40   
max            243.00     510.00                  51.00             350.80   

       total day calls  total day charge  total eve minutes  total eve calls  \
count          3333.00           3333.00            3333.00          3333.00   
mean            100.44             30.56             200.98           100.11   
std              20.07              9.26              50.71            19.92   
min               0.00              0.00               0.00             0.00   
25%              87.00             24.43             166.60            87.00   
50%             101.00             30.50             201.40           100.00   
75%             114.00             36.79             235.30           114.00   
max             165.00             59.64             363.70           170.00   

       total eve charge  total night minutes  total night calls  \
count           3333.00              3333.00            3333.00   
mean              17.08               200.87             100.11   
std                4.31                50.57              19.57   
min                0.00                23.20              33.00   
25%               14.16               167.00              87.00   
50%               17.12               201.20             100.00   
75%               20.00               235.30             113.00   
max               30.91               395.00             175.00   

       total night charge  total intl minutes  total intl calls  \
count             3333.00             3333.00           3333.00   
mean                 9.04               10.24              4.48   
std                  2.28                2.79              2.46   
min                  1.04                0.00              0.00   
25%                  7.52                8.50              3.00   
50%                  9.05               10.30              4.00   
75%                 10.59               12.10              6.00   
max                 17.77               20.00             20.00   

       total intl charge  customer service calls    churn  
count            3333.00                 3333.00  3333.00  
mean                2.76                    1.56     0.14  
std                 0.75                    1.32     0.35  
min                 0.00                    0.00     0.00  
25%                 2.30                    1.00     0.00  
50%                 2.78                    1.00     0.00  
75%                 3.27                    2.00     0.00  
max                 5.40                    9.00     1.00

In order to see statistics on non-numerical features, one has to explicitly indicate data types of interest in the include parameter.

In [21]:
df.describe(include=['object', 'bool'])

state phone number international plan voice mail plan
count   3333         3333               3333            3333
unique    51         3333                  2               2
top       WV     358-1346                 no              no
freq     106            1               3010            2411

For categorical (type object) and boolean (type bool) features we can use the value_counts method. Let's have a look at the distribution of Churn:

In [22]:
df['churn'].value_counts()

0    2850
1     483
Name: churn, dtype: int64

2850 users out of 3333 are loyal; their Churn value is 0. To calculate fractions, pass normalize=True to the value_counts function.    

In [23]:
df['churn'].value_counts(normalize=True)

0    0.86
1    0.14
Name: churn, dtype: float64

#### Check Duplicate valued

We can check if there is any duplicate values in this data frame.

In [24]:
df.duplicated().sum()  # we don't have any duplicate values.

#df.state.duplicated().sum() ### if we need to check duplicate values on column level.

0

In [19]:
df.isnull().sum()  # Will show you null count for each column, but will not count Zeros(0) as null

state                     0
account length            0
area code                 0
phone number              0
international plan        0
voice mail plan           0
number vmail messages     0
total day minutes         0
total day calls           0
total day charge          0
total eve minutes         0
total eve calls           0
total eve charge          0
total night minutes       0
total night calls         0
total night charge        0
total intl minutes        0
total intl calls          0
total intl charge         0
customer service calls    0
churn                     0
dtype: int64

#### Sorting
A DataFrame can be sorted by the value of one of the variables (i.e columns). For example, we can sort by Total day charge (use ascending=False to sort in descending order):

In [25]:
df.sort_values(by=['churn', 'total day charge'], ascending=[True, False]).head()

state  account length  area code phone number international plan  \
688     MN              13        510     338-7120                 no   
2259    NC             210        415     363-7802                 no   
534     LA              67        510     373-6784                 no   
575     SD             114        415     351-7369                 no   
2858    AL             141        510     388-8583                 no   

     voice mail plan  number vmail messages  total day minutes  \
688              yes                     21              315.6   
2259             yes                     31              313.8   
534               no                      0              310.4   
575              yes                     36              309.9   
2858             yes                     28              308.0   

      total day calls  total day charge  ...  total eve calls  \
688               105             53.65  ...               71   
2259               87             53.35  ...              103   
534                97             52.77  ...              123   
575                90             52.68  ...               89   
2858              123             52.36  ...              128   

      total eve charge  total night minutes  total night calls  \
688              17.76                260.1                123   
2259             12.55                192.7                 97   
534               5.65                246.5                 99   
575              17.03                183.5                105   
2858             21.06                152.9                103   

      total night charge  total intl minutes  total intl calls  \
688                11.70                12.1                 3   
2259                8.67                10.1                 7   
534                11.09                 9.2                10   
575                 8.26                14.2                 2   
2858                6.88                 7.4                 3   

      total intl charge  customer service calls  churn  
688                3.27                       3      0  
2259               2.73                       3      0  
534                2.48                       4      0  
575                3.83                       1      0  
2858               2.00                       1      0  

[5 rows x 21 columns]

In [16]:
df.sort_values(by='total day charge', ascending=False).head()


state  account length  area code phone number international plan  \
365     CO             154        415     343-5709                 no   
985     NY              64        415     345-9140                yes   
2594    OH             115        510     348-1163                yes   
156     OH              83        415     370-9116                 no   
605     MO             112        415     373-2053                 no   

     voice mail plan  number vmail messages  total day minutes  \
365               no                      0              350.8   
985               no                      0              346.8   
2594              no                      0              345.3   
156               no                      0              337.4   
605               no                      0              335.5   

      total day calls  total day charge  ...  total eve calls  \
365                75             59.64  ...               94   
985                55             58.96  ...               79   
2594               81             58.70  ...              106   
156               120             57.36  ...              116   
605                77             57.04  ...              109   

      total eve charge  total night minutes  total night calls  \
365              18.40                253.9                100   
985              21.21                275.4                102   
2594             17.29                217.5                107   
156              19.33                153.9                114   
605              18.06                265.0                132   

      total night charge  total intl minutes  total intl calls  \
365                11.43                10.1                 9   
985                12.39                13.3                 9   
2594                9.79                11.8                 8   
156                 6.93                15.8                 7   
605                11.93                12.7                 8   

      total intl charge  customer service calls  churn  
365                2.73                       1      1  
985                3.59                       1      1  
2594               3.19                       1      1  
156                4.27                       0      1  
605                3.43                       2      1  

[5 rows x 21 columns]

In [26]:
df['churn'].mean()

0.14491449144914492

14.5% is actually quite bad for a company; such a churn rate can make the company go bankrupt


#### What are average values of numerical features for churned users?

In [28]:
df[df['churn'] == 1].mean()

account length            102.66
area code                 437.82
number vmail messages       5.12
total day minutes         206.91
total day calls           101.34
total day charge           35.18
total eve minutes         212.41
total eve calls           100.56
total eve charge           18.05
total night minutes       205.23
total night calls         100.40
total night charge          9.24
total intl minutes         10.70
total intl calls            4.16
total intl charge           2.89
customer service calls      2.23
churn                       1.00
dtype: float64

DataFrames can be indexed by column name (label) or row name (index) or by the serial number of a row. The loc method is used for indexing by name, while iloc() is used for indexing by number.

In the first case below, we say "give us the values of the rows with index from 0 to 5 (inclusive) and columns labeled from State to Area code (inclusive)". In the second case, we say "give us the values of the first five rows in the first three columns" (as in a typical Python slice: the maximal value is not included).

In [38]:
df.loc[0:5, 'state':'area code']

state  account length  area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [39]:
df.iloc[0:5, 0:3]

state  account length  area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

If we need the first or the last line of the data frame, we can use the df[:1] or df[-1:] construct:

In [40]:
df[-1:]

state  account length  area code phone number international plan  \
3332    TN              74        415     400-4344                 no   

     voice mail plan  number vmail messages  total day minutes  \
3332             yes                     25              234.4   

      total day calls  total day charge  ...  total eve calls  \
3332              113             39.85  ...               82   

      total eve charge  total night minutes  total night calls  \
3332              22.6                241.4                 77   

      total night charge  total intl minutes  total intl calls  \
3332               10.86                13.7                 4   

      total intl charge  customer service calls  churn  
3332                3.7                       0      0  

[1 rows x 21 columns]

### Grouping
In general, grouping data in Pandas works as follows:

df.groupby(by=grouping_columns)[columns_to_show].function()
First, the groupby method divides the grouping_columns by their values. They become a new index in the resulting dataframe.
Then, columns of interest are selected (columns_to_show). If columns_to_show is not included, all non groupby clauses will be included.
Finally, one or several functions are applied to the obtained groups per selected columns.
Here is an example where we group the data according to the values of the Churn variable and display statistics of three columns in each group:

Let’s do the same thing, but slightly differently by passing a list of functions to agg():

In [41]:
columns_to_show = ['total day minutes', 'total eve minutes', 
                   'total night minutes']

df.groupby(['churn'])[columns_to_show].agg([np.mean, np.std, np.min, 
                                            np.max])

total day minutes                                   total eve minutes  \
                  count    mean    std  min    50%    max             count   
churn                                                                         
0                2850.0  175.18  50.18  0.0  177.2  315.6            2850.0   
1                 483.0  206.91  69.00  0.0  217.6  350.8             483.0   

                                         total night minutes                 \
         mean    std   min    50%    max               count    mean    std   
churn                                                                         
0      199.04  50.29   0.0  199.6  361.8              2850.0  200.13  51.11   
1      212.41  51.73  70.9  211.3  363.7               483.0  205.23  47.13   

                            
        min     50%    max  
churn                       
0      23.2  200.25  395.0  
1      47.4  204.80  354.9

### Summary tables
Suppose we want to see how the observations in our sample are distributed in the context of two variables - Churn and International plan. To do so, we can build a contingency table using the crosstab method:

In [43]:
pd.crosstab(df['churn'], df['international plan'])

international plan    no  yes
churn                        
0                   2664  186
1                    346  137